### PR TITLE
Update data importer to support CJK

### LINKED ISSUE #741: Implement convertion between Chinese Traditional and Simplified
Chinese script comes in traditional and simplified variety. Most big translation vendors support both. Converting traditional to simplified (and vice versa) can be easily achieved through hanzi-conv https://pypi.org/project/hanziconv/0.3/ . There might be a very small information loss when converting simplified to traditional, but it should be fine in 99.9% of the cases. Some datasets such as ted talks come in traditional, so they should be converted before using.

See also this comment by Jaume:
https://github.com/mozilla/firefox-translations-training/issues/45#issuecomment-1036191497

### LINKED ISSUE #752: Support data import for CJK
Make sure we map language codes correctly. Maybe there are some other things to adjust.

### PR DESCRIPTION
**Pull request by eu9ene**
The goal is to treat the language code zh as Mandarin Chinese in Simplified script for now.

- convert all Chinese to Simplified (support for Traditional will be handled separately in #896)
- add cjk augmentation modifier

closes #752 
closes #741

### CODE REVIEW

File: pipeline/data/download-mono.py [hunk 1]
```diff
@@ -119,6 +121,17 @@ def main(args_list: Optional[list[str]] = None) -> None:
         ):
             outfile.write(line)
 
+    # TODO: convert everything to Chinese simplified for now
```
Comment (gregtatum) [inline pipeline/data/download-mono.py:124]: Suggestion (optional): I believe this change isn't covered by a test.

**Review by ZJaume — APPROVED** [review]

File: pipeline/data/dataset_importer.py [hunk 2]
```diff
@@ -79,18 +77,28 @@ def get_typos_probs() -> Dict[str, float]:
             PlaceholderTagModifier(NOISE_MIX_PROB, augment=1),
         ]
     ),
+    "aug-mix-cjk": lambda: CompositeModifier(
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:80]: Those are documented in https://github.com/mozilla/translations/blob/main/docs/opus-trainer.md#supported-modifiers. I'm adding the new one there too.

File: tests/test_data_importer.py [hunk 3]
```diff
@@ -332,9 +327,53 @@ def test_augmentation_mix(data_dir):
...
+    len_noise_src = len(aug_src) - len(src)
+    len_noise_trg = len(aug_trg) - len(trg)
+    # check noise rate
+    for noise, original in [(len_noise_src, len(src)), (len_noise_trg, len(trg))]:
+        noise_rate = noise / original
+        assert noise_rate > AUG_MIN_RATE
+        assert noise_rate < AUG_MAX_RATE
+
+    # check augmentation rate without noise
+    for aug, original in [(aug_src, src), (aug_trg, trg)]:
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply tests/test_data_importer.py:354]: It's a git diff issue. I parametrized this test now, so there's no duplication.

File: tests/test_data_importer.py [hunk 4]
```diff
@@ -111,10 +193,10 @@ def test_basic_corpus_import(importer, dataset, data_dir):
...
 mono_params = [
-    ("news-crawl", "en", "news_2021",                    [0, 1, 4, 6, 3, 7, 5, 2]),
-    ("news-crawl", "ru", "news_2021",                    [0, 1, 4, 6, 3, 7, 5, 2]),
-    ("url",        "en", "gcp_pytest-dataset_en_cdd0d7", [2, 1, 5, 4, 0, 7, 6, 3]),
-    ("url",        "ru", "gcp_pytest-dataset_ru_be3263", [5, 4, 2, 0, 7, 1, 3, 6]),
+    ("news-crawl", "en", "news_2021", [0, 1, 4, 6, 3, 7, 5, 2]),
+    ("news-crawl", "ru", "news_2021", [0, 1, 4, 6, 3, 7, 5, 2]),
+    ("url", "en", "gcp_pytest-dataset_en_cdd0d7", [2, 1, 5, 4, 0, 7, 6, 3]),
+    ("url", "ru", "gcp_pytest-dataset_ru_be3263", [5, 4, 2, 0, 7, 1, 3, 6]),
 ]  # fmt: skip
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply tests/test_data_importer.py:200]: it was an accident, fixed

File: pipeline/data/dataset_importer.py [hunk 5]
```diff
@@ -221,10 +238,28 @@ def run_import(type: str, dataset: str, output_prefix: str):
...
+            if lang == "zh":
+                print("Converting the output file to Chinese Simplified")
+                chinese_converter = ChineseConverter()
+                count = chinese_converter.convert_file(
+                    f"{output_prefix}.{lang}.zst",
+                    f"{output_prefix}.converted.{lang}.zst",
+                    ChineseType.simplified,
+                )
+                shutil.move(f"{output_prefix}.converted.{lang}.zst", f"{output_prefix}.{lang}.zst")
+                print(f"Converted {count} lines to Chinese Simplified")
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:258]: cool, thanks for the code, it outputs this nice json now:

```json
{
  "script": "ChineseType.simplified",
  "script_conversion": {
    "description": "How many sentences in the dataset were converted to simplified",
    "converted": 0,
    "visited": 1012
  }
}
```

**Comment by eu9ene** [comment]
@gregtatum I addressed all your comments so it's ready for another pass

File: pipeline/data/cjk.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+from enum import Flag
```
Comment (gregtatum) [inline pipeline/data/cjk.py:1]: This needs a document-level comment explaining what the file does.

File: pipeline/data/dataset_importer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/data/dataset_importer.py:83]: Please add an inline link to this documentation so it's cross-referenced.

**Review by gregtatum — APPROVED** [review]

File: pipeline/data/dataset_importer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:83]: Look at the full diff, I see the modifier documentation there. Or do you mean something else should be added?

[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:84]: I added the inline reference. I need to merge this one to unblock other PRs. Let's follow up with another PR if something else is missing.

[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/data/dataset_importer.py:84]: Yeah I think it's not clear the difference between `aug-mix` and `aug-mix-cjk` that they are mutually exclusive options.

**Pull request merged**